### PR TITLE
Load Java 17 parser in a classloader that avoids module security

### PR DESCRIPTION
**Pull request by tkvangorder**

### CODE REVIEW

File: rewrite-java-17/src/main/java/org/openrewrite/java/Java17Parser.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+        }
+
+        @Override
+        public Java17Parser build() {
+            lazyInitClassLoaders();
+
+            try {
+                // need to reverse this parent/child relationship
+                Class<?> parserImplementation = Class.forName("org.openrewrite.java.isolated.IsolatedJava17Parser", true, moduleClassLoader);
```
Comment (jkschneider) [inline rewrite-java-17/src/main/java/org/openrewrite/java/Java17Parser.java:89]: Maybe follow the same naming as the reflection loaded Java8 parser? There we called it `ReloadableJava8Parser` and I think it was in the `.internal` package?

Comment (tkvangorder) [reply rewrite-java-17/src/main/java/org/openrewrite/java/Java17Parser.java]: I renamed the classes by prefixing them with "Reloadable". I left the classes in `org.openrewrite.java.isolated` as `rewrite-java` has classes in `org.openrewrite.java.internal` and we do not want those loaded by the new class loader. 

Right now the class loader uses the package name to determine which classes are loaded vs loaded by the parent. This was easier than enumerating specific classes (and nested classes)

File: rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java [hunk 2]
```diff
@@ -15,445 +15,188 @@
...
+     * default behavior is to throw a security exception when attempting to use these internal classes. This classloader
+     * circumvents these security restrictions by isolating Rewrite's Java 17 parser implementation classes and then
+     * loading any of the internal classes directly from the .jmod files.
+     *
+     * NOTE: Any classes in the package "org.openrewrite.java.isolated" will be loaded into this isolated classloader.
+     */
+    private static class UnrestrictedModuleClassLoader extends ClassLoader {
+
+        static {
+            ClassLoader.registerAsParallelCapable();
```
Comment (tkvangorder) [inline rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java:125]: Note, I am registering this classloader as "parallel capable" and then using the same pattern used in ClassLoader for synchronizing the loadClass method. See https://github.com/openrewrite/rewrite/pull/1797/files#diff-8bc6211c47504aaff7724ac3f94894b07ccf54ac7e77a40fbe8e7e6a7bd48177R145

Without this, I was getting failures (only on the CI builds)

File: rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java [hunk 3]
```diff
@@ -15,445 +15,188 @@
...
-import static java.util.Collections.emptyList;
-import static java.util.stream.Collectors.toList;
+import java.util.Collection;
+import java.util.List;
 
-/**
- * This parser is NOT thread-safe, as the OpenJDK parser maintains in-memory caches in static state.
- */
-@NonNullApi
 public class Java11Parser implements JavaParser {
```
Comment (sambsnyd) [inline rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java:38]: Do these changes enhance the behavior of `Java11Parser`? Is the main reason to change it for consistency with the Java 8 / 17 implementations?

**Comment by jkschneider** [comment]
I'm kind of confused about why we made Java 11 parser reloadable. In Java 11, I thought it was fine to add module exports at compilation time, and then there was no runtime effect.

File: rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java (hunk 3 above)
Comment (tkvangorder) [reply rewrite-java-11/src/main/java/org/openrewrite/java/Java11Parser.java:38]: Doing this to the Java11Parser means that you no longer have to add module exports at runtime (ever). So the only time this is a requirement is when we are compiling. This means that all three parsers "just work" at runtime, regardless of which JVM is hosting them.

**Comment by tkvangorder** [comment]
Since Gradle is now running in a Java 17 JVM now, we would either have to add exports to Gradle's testing tasks (to support running Java 11 parser without modification). And if we did add exports to the runtime, I would expect the Java 17 parser to work as well. So rather than having a mixed bag (one parser that circumvents jigsaw security and one parser that doesn't), I just made both parsers use the security circumventing loader.

**Pull request merged**